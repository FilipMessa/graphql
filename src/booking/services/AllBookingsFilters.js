// @flow

import idx from 'idx';
import { DateTime } from 'luxon';

import type { BookingsItem } from '../Booking';
import type { Leg } from '../../flight/Flight';
import { CoveredBy } from '../../flight/types/enums/CoveredBy';

export function groupBookings(bookings: $ReadOnlyArray<BookingsItem>) {
  return bookings.reduce(
    (acc, curVal) => {
      if (isPastBooking(curVal) === true) {
        acc.past.push(curVal);
      } else {
        acc.future.push(curVal);
      }
      return acc;
    },
    {
      future: [],
      past: [],
    },
  );
}

export function filterOnlyBookings(
  timeFrame: 'future' | 'past',
  bookings: $ReadOnlyArray<BookingsItem>,
) {
  const bookingGroups = groupBookings(bookings);

  if (timeFrame === 'future') {
    return bookingGroups.future;
  }
  return bookingGroups.past;
}

/**
 * Past booking is booking where every arrival date of every leg is in past
 * (considering local time).
 */
export function isPastBooking(singleBooking: BookingsItem) {
  const when = singleBooking.arrival.when;
  if (!when) {
    return true;
  }

  return (
    DateTime.fromJSDate(when.utc, {
      zone: 'utc',
    })
      .diffNow('milliseconds')
      .toObject().milliseconds < 0
  );
}

export function findUpcomingLeg(
  legs: Leg[],
  guarantee: ?$Values<typeof CoveredBy>,
) {
  return [...legs]
    .filter(leg => {
      const date = idx(leg, _ => _.arrival.when.utc);
      const isFuture = date
        ? DateTime.fromJSDate(date, { zone: 'utc' }).diffNow('milliseconds')
            .milliseconds > 0
        : false;

      return guarantee ? leg.guarantee === guarantee && isFuture : isFuture;
    })
    .sort((legA, legB) => {
      const dateA = idx(legA, _ => _.arrival.when.utc);
      const dateB = idx(legB, _ => _.arrival.when.utc);

      return dateA && dateB ? dateA - dateB : 0;
    })
    .shift();
}
