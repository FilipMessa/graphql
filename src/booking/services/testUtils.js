// @flow

export const createLeg = (milliseconds: number) => ({
  arrival: {
    when: {
      utc: new Date(milliseconds),
    },
  },
});
