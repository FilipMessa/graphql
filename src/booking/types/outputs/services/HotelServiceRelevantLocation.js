// @flow

import { GraphQLObjectType } from 'graphql';
import idx from 'idx';

import Location from '../../../../location/types/outputs/Location';
import type { GraphqlContextType } from '../../../../common/services/GraphqlContext';
import HotelCityGraphqQLType from '../../../../hotel/types/outputs/HotelCity';

export default new GraphQLObjectType({
  name: 'HotelServiceRelevantLocation',
  fields: {
    location: {
      type: Location,
      resolve: async (
        ancestor: string,
        _: mixed,
        { dataLoader }: GraphqlContextType,
      ) => {
        const location = await dataLoader.location.loadById(ancestor);
        return location;
      },
    },
    hotelCity: {
      type: HotelCityGraphqQLType,
      resolve: async (
        ancestor: string,
        _: mixed,
        { dataLoader, locale }: GraphqlContextType,
      ) => {
        const location = await dataLoader.location.loadById(ancestor);
        let hotelCities = [];
        const cityName = idx(location, _ => _.city.name);

        if (cityName) {
          hotelCities = await dataLoader.hotel.cities.loadByPrefix(
            cityName,
            locale.language,
          );
        }

        const lat = idx(location, _ => _.location.lat);
        const lng = idx(location, _ => _.location.lng);
        if (hotelCities.length === 0 && lat != null && lng != null) {
          /**
           * If we don't find any city quering by name, let's fall back to the coordinates of the airport
           * Note that this is most likely not going to be the destination city
           */
          hotelCities = await dataLoader.hotel.cities.loadByLatLng(lat, lng);
        }

        return hotelCities[0];
      },
    },
  },
});
