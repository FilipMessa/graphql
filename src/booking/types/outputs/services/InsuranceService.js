// @flow

import { GraphQLObjectType, GraphQLList } from 'graphql';

import GraphQLPassenger from '../Passenger';

export default new GraphQLObjectType({
  name: 'InsuranceService',
  fields: {
    passengers: {
      type: GraphQLList(GraphQLPassenger),
      description: 'Returns passengers allowed to buy a travel insurance.',
    },
  },
});
