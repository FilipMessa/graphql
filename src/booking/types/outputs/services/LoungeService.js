// @flow

import { GraphQLObjectType, GraphQLString, GraphQLList } from 'graphql';
import Location from '../../../../location/types/outputs/Location';
import LoungeWhiteLabelURLResolver from '../../../resolvers/LoungeWhitelabelURL';
import type { GraphqlContextType } from '../../../../common/services/GraphqlContext';

type LoungeServiceInput = {|
  +iataCodes: Set<string>,
  +departureTime: Date,
|};

type LoungeServiceOutput = {|
  +iataCode: string,
  +departureTime: Date,
|};

export default new GraphQLObjectType({
  name: 'LoungeService',
  fields: {
    relevantAirports: {
      type: GraphQLList(
        new GraphQLObjectType({
          name: 'LoungeServiceRelevantAirports',
          fields: {
            whitelabelURL: {
              type: GraphQLString,
              resolve: async ({
                iataCode,
                departureTime,
              }: LoungeServiceOutput) => {
                return LoungeWhiteLabelURLResolver(iataCode, departureTime);
              },
            },

            location: {
              type: Location,
              resolve: (
                { iataCode }: LoungeServiceOutput,
                args: void,
                context: GraphqlContextType,
              ) => {
                return context.dataLoader.location.load(iataCode);
              },
            },
          },
        }),
      ),
      resolve: ({
        iataCodes,
        departureTime,
      }: LoungeServiceInput): $ReadOnlyArray<LoungeServiceOutput> => {
        return Array.from(iataCodes).map(iataCode => ({
          iataCode,
          departureTime,
        }));
      },
    },
  },
});
