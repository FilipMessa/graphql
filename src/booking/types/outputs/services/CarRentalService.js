// @flow

import { GraphQLObjectType, GraphQLList } from 'graphql';

import CarRentalServiceRelevantCities, {
  type CarRentalServiceRelevantCitiesType,
} from './CarRentalServiceRelevantCities';
import { type GraphqlContextType } from '../../../../common/services/GraphqlContext';

export type CarRentalServiceType = Map<
  string,
  {|
    +pickup: Date,
    +dropoff: Date,
  |},
>;

export default new GraphQLObjectType({
  name: 'CarRentalService',
  fields: {
    relevantCities: {
      type: GraphQLList(CarRentalServiceRelevantCities),
      resolve: async (
        ancestor: CarRentalServiceType,
        args: {},
        context: GraphqlContextType,
      ): Promise<CarRentalServiceRelevantCitiesType[]> => {
        const locations = await Promise.all(
          Array.from(ancestor.keys()).map(IATA =>
            context.dataLoader.location.loadById(IATA),
          ),
        );

        return locations.map(location => {
          const ancestorValue = ancestor.get(location.locationId) || {
            pickup: new Date(),
            dropoff: new Date(),
          };

          return {
            location,
            pickup: ancestorValue.pickup,
            dropoff: ancestorValue.dropoff,
          };
        });
      },
    },
  },
});
