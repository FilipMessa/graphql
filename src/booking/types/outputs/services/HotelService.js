// @flow

import { GraphQLObjectType, GraphQLList } from 'graphql';

import HotelServiceRelevantLocation from './HotelServiceRelevantLocation';

type AncestorType = {|
  +relevantLocationCodes: $ReadOnlyArray<string>,
|};

export default new GraphQLObjectType({
  name: 'HotelService',
  fields: {
    relevantLocations: {
      type: GraphQLList(HotelServiceRelevantLocation),
      resolve: ({ relevantLocationCodes }: AncestorType) =>
        relevantLocationCodes,
    },
  },
});
