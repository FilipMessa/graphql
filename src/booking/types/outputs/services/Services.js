// @flow

import { GraphQLObjectType } from 'graphql';
import { DateTime } from 'luxon';

import InsuranceService from './InsuranceService';
import type { BookingsItem } from '../../../Booking';

type AncestorType = {|
  +booking: BookingsItem,
|};

export default new GraphQLObjectType({
  name: 'Services',
  fields: {
    insurance: {
      type: InsuranceService,
      description: 'Travel insurance - returns "null" if not available.',
      resolve: async ({ booking }: AncestorType) => {
        // booking must be CONFIRMED
        if (booking.status !== 'confirmed') {
          return null;
        }

        if (!booking.departure.when) {
          return null;
        }

        // departure must be at least 48 hours from now
        const diff = DateTime.fromJSDate(booking.departure.when.utc)
          .diff(DateTime.fromJSDate(new Date()), 'hours')
          .toObject().hours;

        if (diff < 48) {
          return null;
        }

        // we do not provide insurance for US nationality (need to check all PAX)
        const validPax = booking.passengers.filter(
          pap => pap.nationality !== 'us',
        );
        if (validPax.length === 0) {
          return null;
        }

        return {
          passengers: validPax,
        };
      },
    },
  },
});
