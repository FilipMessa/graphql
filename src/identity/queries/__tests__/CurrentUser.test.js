// @flow

import { graphql, RestApiMock } from '../../../common/services/TestingTools';
import config from '../../../../config/application';
import CurrentUserDataset from '../../datasets/currentUser.json';
import CurrentUser from '../CurrentUser';

beforeEach(() => {
  RestApiMock.onPost(config.restApiEndpoint.identity).replyWithData(
    CurrentUserDataset,
  );
});

describe('current user query', () => {
  it('should be of Identity type', () => {
    expect(CurrentUser.type.toString()).toBe('Identity');
  });

  it('should return valid fields', async () => {
    const query = `{
      currentUser {
        id
        lastName
        firstName
        fullName
        email
        emailVerified
        login
      }
    }`;
    expect(await graphql(query)).toMatchSnapshot();
  });
});
