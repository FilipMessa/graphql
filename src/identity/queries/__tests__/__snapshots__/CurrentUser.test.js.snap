// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`current user query should return valid fields 1`] = `
Object {
  "data": Object {
    "currentUser": Object {
      "email": "kiwicomtester@gmail.com",
      "emailVerified": true,
      "firstName": "Kiwicom",
      "fullName": "Kiwicom Tester",
      "id": "aWRlbnRpdHk6UzVUU1kwTnVzVk9zUVJUeVJwTWVfRw==",
      "lastName": "Tester",
      "login": "kiwicomtester@gmail.com",
    },
  },
}
`;
